pywr-core/src/timestep.rs

```diff
@@ -1,79 +1,169 @@
     pub index: TimestepIndex,
-    pub duration: Duration,
+    pub duration: PywrDuration,
 }
 
 impl Timestep {
-    pub fn new(date: Date, index: TimestepIndex, duration: Duration) -> Self {
+    pub fn new(date: NaiveDateTime, index: TimestepIndex, duration: PywrDuration) -> Self {
         Self { date, index, duration }
     }
 
     pub fn is_first(&self) -> bool {
         self.index == 0
     }
 
-    // pub fn parse_from_str(date: &str, fmt: &str, index: TimestepIndex, timestep: i64) -> Result<Self, PywrError> {
-    //     Ok(Self {
-    //         date: Date::parse_from_str(date, fmt)?,
-    //         index,
-    //         duration: Duration::days(timestep),
-    //     })
-    // }
-
     pub(crate) fn days(&self) -> f64 {
-        self.duration.as_seconds_f64() / 3600.0 / 24.0
+        self.duration.fractional_days()
     }
 }
 
-impl Add<Duration> for Timestep {
+impl Add<PywrDuration> for Timestep {
     type Output = Timestep;
 
-    fn add(self, other: Duration) -> Self {
+    fn add(self, other: PywrDuration) -> Self {
         Self {
-            date: self.date + other,
+            date: self.date + other.0,
             index: self.index + 1,
             duration: other,
         }
     }
 }
 
+#[derive(Debug)]
+pub enum TimestepDuration {
+    Days(i64),
+    Frequency(String),
+}
+
 #[derive(Debug)]
 pub struct Timestepper {
-    start: Date,
-    end: Date,
-    timestep: Duration,
+    start: NaiveDateTime,
+    end: NaiveDateTime,
+    timestep: TimestepDuration,
 }
 
 impl Timestepper {
-    pub fn new(start: Date, end: Date, timestep: i64) -> Self {
-        Self {
-            start,
-            end,
-            timestep: Duration::days(timestep),
-        }
+    pub fn new(start: NaiveDateTime, end: NaiveDateTime, timestep: TimestepDuration) -> Self {
+        Self { start, end, timestep }
     }
 
     /// Create a vector of `Timestep`s between the start and end dates at the given duration.
-    fn timesteps(&self) -> Vec<Timestep> {
+    fn timesteps(&self) -> Result<Vec<Timestep>, PywrError> {
+        match self.timestep {
+            TimestepDuration::Days(days) => Ok(self.generate_timesteps_from_days(days)),
+            TimestepDuration::Frequency(ref frequency) => self.generate_timesteps_from_frequency(frequency.clone()),
+        }
+    }
+
+    fn generate_timesteps_from_days(&self, days: i64) -> Vec<Timestep> {
         let mut timesteps: Vec<Timestep> = Vec::new();
-        let mut current = Timestep::new(self.start, 0, self.timestep);
+        let duration = PywrDuration::days(days);
+        let mut current = Timestep::new(self.start, 0, duration);
 
         while current.date <= self.end {
-            let next = current + self.timestep;
+            let next = current + duration;
             timesteps.push(current);
             current = next;
         }
         timesteps
     }
+
+    fn generate_timesteps_from_frequency(&self, frequency: String) -> Result<Vec<Timestep>, PywrError> {
```

Comment:
Can we add a docstring here. Also it would be more idiomatic to use `frequency: &str`.